pkg/dao/rpm.go

```diff
@@ -0,0 +1,203 @@
+package dao
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/content-services/content-sources-backend/pkg/api"
+	"github.com/content-services/content-sources-backend/pkg/models"
+	"gorm.io/gorm"
+)
+
+type rpmDaoImpl struct {
+	db *gorm.DB
+}
+
+func GetRpmDao(db *gorm.DB) RpmDao {
+	return rpmDaoImpl{
+		db: db,
+	}
+}
+
+func (r rpmDaoImpl) isOwnedRepository(orgID string, repoUUID string) (bool, error) {
+	var repoConfigs []models.RepositoryConfiguration
+	var count int64
+	if err := r.db.
+		Where("org_id = ? and repository_uuid = ?", orgID, repoUUID).
+		Find(&repoConfigs).
+		Count(&count).
+		Error; err != nil {
+		return false, err
+	}
+	if count == 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// Create a record in rpms table, and the relation between it
+// and the Repository record in the repositories_rpms table.
+// Before create the record, it checks that the provided
+// Repository register is related with a RepositoryConfiguration
+// that belong to the indicated organization.
+// orgID It is used to check the repository which we are trying
+// to add the Rpm record belongs to the indicated organization.
+// repo The repository record that the new rpm record will
+// belong to.
+// newRpm The Rpm record to be created into the database.
+// Return error if something goes wrong, else nil.
+func (r rpmDaoImpl) Create(orgID string, repo *models.Repository, newRpm *models.Rpm) error {
+	// Check arguments
+	if repo == nil {
+		return fmt.Errorf("repo can not be nil")
+	}
+	if newRpm == nil {
+		return fmt.Errorf("newRpm can not be nil")
+	}
+
+	if ok, err := r.isOwnedRepository(orgID, repo.UUID); !ok {
+		if err != nil {
+			return DBErrorToApi(err)
+		}
+		return fmt.Errorf("repository_uuid = %s is not owned", repo.UUID)
+	}
+
+	// Add Rpm record
+	if err := r.db.Create(newRpm).Error; err != nil {
+		return DBErrorToApi(err)
+	}
+
+	// Add to repositories_rpms the entry to relate
+	// the rpm with the repository it belongs to
+	var repositories_rpms []map[string]interface{} = []map[string]interface{}{
+		{
+			"repository_uuid": repo.UUID,
+			"rpm_uuid":        newRpm.UUID,
+		},
+	}
+	if err := r.db.Table(models.TableNameRpmsRepositories).Create(&repositories_rpms).Error; err != nil {
```

Comment:
why not have a separate model for the `repositories_rpms` table?

Reply:
+1 to this

Reply:
thanks! Indeed I defined at some point the type struct, but I removed it because I didn't use it.

Checking again the code, I don't like how I have made this; I am hard-coding the name of the fields into the map[string] which make it more risky to future changes; using the model here future changes are detected in compilation time instead of runtime with the tests and it improves the maintenance.

Updated